src/client.js

```diff
@@ -13,133 +14,385 @@ export function getAddress(system) {
+                    imageLoadChannel.addEventListener("close", function (e, options) {
+                        if (options.problem)
+                            reject(new Error(options.problem));
+                        else if (imageLoadChannelResponse.status >= 200 && imageLoadChannelResponse.status <= 299)
+                            resolve();
+                        else
+                            reject(new Error(String.fromCharCode.apply(null, imageLoadChannelData)));
+                    })
+                );
+            } finally {
+                await cockpit.script(`rm "${imageFile}"`, { host: targetHost });
+            }
+
+            // Inspect the old image and try to tag the new one to match it
+            callbacks.imageTagging();
+            const oldImageInfo = await inspectImage(system, containerInfo.Image, host);
+            for (const tag of oldImageInfo.RepoTags) {
+                const i = tag.lastIndexOf(":");
+                try {
+                    await tagImage(system, containerInfo.Image, tag.substring(0, i), tag.substring(i + 1, tag.length),
+                                   targetHost);
+                } catch (error) {
+                    // Ignore
+                }
+            }
+        } else {
+            throw error;
+        }
+    }
+
+    let containerCreated = false;
+    try {
+        callbacks.containerInspect();
+        await inspectContainer(system, containerInfo.Name, targetHost);
+    } catch (error) {
+        if (error.status === 404) {
+            callbacks.containerCreate();
+            const dummyConfig = {
+                name: containerInfo.Name,
+                image: containerInfo.Image,
+                terminal: true,
+                command: containerInfo.Args
+            };
+            await createContainer(system, dummyConfig, targetHost);
+            containerCreated = true;
+        } else {
+            throw error;
+        }
+    }
+
+    // Checkpoint the selected container
+    callbacks.containerExport();
+    const containerFile = (await cockpit.script("mktemp -p /var/tmp", { host: targetHost })).slice(0, -1);
+    const containerSaveChannel = cockpit.channel({
+        payload: "fsreplace1",
+        path: containerFile,
+        binary: true,
+        host: targetHost
+    });
+    await containerSaveChannel.wait();
+    const checkpointArgs = {
+        keep: args.keep,
+        leaveRunning: args.leaveRunning,
+        tcpEstablished: args.tcpEstablished,
+        ignoreRootFS: args.ignoreRootFS,
+        ignoreStaticIP: args.ignoreStaticIP,
+        ignoreStaticMAC: args.ignoreStaticMAC
+    };
+    await exportContainer(system, id, checkpointArgs, host, containerSaveChannel.id);
+    containerSaveChannel.control({ command: "done" });
+    await new Promise(resolve => containerSaveChannel.addEventListener("close", resolve));
+
+    // Restore it into the created dummy container
+    callbacks.containerImport();
```

Comment:
Ok, I don't get this dummy container thing. According to https://criu.org/Podman#Container_Live_Migration you export into a file and then restore from a file. No need for any tmp containers.

Reply:
That is true when using the Podman CLI, but with the REST API it's different - the container has to exist (see https://docs.podman.io/en/latest/_static/api.html#operation/libpodRestoreContainer and https://github.com/containers/podman/blob/master/pkg/api/handlers/libpod/containers.go).